server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -1814,4 +1836,114 @@ List<SearchShardIterator> getLocalShardsIterator(
+        private final SearchTask task;
+        private final ActionListener<SearchResponse> listener;
+        private final CCSUsage.Builder usageBuilder;
+
+        SearchResponseActionListener(SearchTask task, ActionListener<SearchResponse> listener) {
+            this.task = task;
+            this.listener = listener;
+            usageBuilder = new CCSUsage.Builder();
+        }
+
+        /**
+         * Should we collect telemetry for this search?
+         */
+        private boolean collectTelemetry() {
+            return CCS_TELEMETRY_FEATURE_FLAG.isEnabled() && usageBuilder.getRemotesCount() > 0;
+        }
+
+        public void setRemotes(int count) {
+            usageBuilder.setRemotesCount(count);
+        }
+
+        @Override
+        public void setFeature(String feature) {
+            usageBuilder.setFeature(feature);
+        }
+
+        @Override
+        public void setClient(String client) {
+            usageBuilder.setClient(client);
+        }
+
+        @Override
+        public void onResponse(SearchResponse searchResponse) {
+            try {
+                searchResponseMetrics.recordTookTime(searchResponse.getTookInMillis());
+                SearchResponseMetrics.ResponseCountTotalStatus responseCountTotalStatus =
+                    SearchResponseMetrics.ResponseCountTotalStatus.SUCCESS;
+                if (searchResponse.getShardFailures() != null && searchResponse.getShardFailures().length > 0) {
+                    // Deduplicate failures by exception message and index
+                    ShardOperationFailedException[] groupedFailures = ExceptionsHelper.groupBy(searchResponse.getShardFailures());
+                    for (ShardOperationFailedException f : groupedFailures) {
+                        boolean causeHas500Status = false;
+                        if (f.getCause() != null) {
+                            causeHas500Status = ExceptionsHelper.status(f.getCause()).getStatus() >= 500;
+                        }
+                        if ((f.status().getStatus() >= 500 || causeHas500Status)
+                            && ExceptionsHelper.isNodeOrShardUnavailableTypeException(f.getCause()) == false) {
+                            logger.warn("TransportSearchAction shard failure (partial results response)", f);
+                            responseCountTotalStatus = SearchResponseMetrics.ResponseCountTotalStatus.PARTIAL_FAILURE;
+                        }
+                    }
+                }
+                // increment after the delegated onResponse to ensure we don't
+                // record both a success and a failure if there is an exception
+                searchResponseMetrics.incrementResponseCount(responseCountTotalStatus);
+
+                if (collectTelemetry()) {
+                    extractCCSTelemetry(searchResponse);
+                    recordTelemetry();
+                }
+                // TODO: should this be last?
```

Comment:
Ok I'll look into it next week.

Reply:
I think I understand now what's going on there. If we increment the counter and collect the telemetry, and then `listener.onResponse(searchResponse);` throws, it will go to `onFailure`, which will increment the counter and try to do telemetry for the second time, which is not good.

Reply:
I am not sure why this try/catch exists at all. It was added in https://github.com/elastic/elasticsearch/commit/9685a5d32cc5be97d5eca77277d171cf232e9904 but I am not sure it's how it should be working...